File: frontend/lib/engine/exports.ts
import { utils, write } from "xlsx";
import type { ResultRecord } from "../types";
import type { ExportType, ExportFormat } from "../services/contracts";
export const exportColumns: Record<ExportType, string[]> = {
  valid: ["email"],
  corrected: ["original_email", "corrected_email", "reason"],
  removed: ["original_email", "category", "reason", "domain"],
  review: ["original_email", "final_email", "reason", "domain"],
  clean: ["email"],
  full: [
    "original_email",
    "normalized_email",
    "final_email",
    "status",
    "category",
    "reason",
    "domain",
    "was_corrected",
    "correction_type",
    "row_number",
  ],
};
const text = (value: unknown) =>
  typeof value === "boolean" ? (value ? "True" : "False") : String(value);
export function exportRow(
  record: ResultRecord,
  kind: ExportType,
): string[] | null {
  if (kind === "clean" || kind === "valid")
    return record.status === "VALID" ||
      (kind === "clean" && record.status === "CORRECTED")
      ? [record.final_email]
      : null;
  if (kind !== "full" && record.status.toLowerCase() !== kind) return null;
  return exportColumns[kind].map((key) =>
    text(
      record[
        (key === "corrected_email" ? "final_email" : key) as keyof ResultRecord
      ],
    ),
  );
}
export function createExport(
  records: ResultRecord[],
  kind: ExportType,
  format: ExportFormat,
): Blob {
  if (!exportColumns[kind] || !["csv", "xlsx"].includes(format))
    throw new Error("Unsupported export type or format");
  const rows: string[][] = [exportColumns[kind]],
    seen = new Set<string>();
  for (const record of records) {
    const row = exportRow(record, kind);
    if (!row) continue;
    if (kind === "clean") {
      if (seen.has(row[0])) continue;
      seen.add(row[0]);
    }
    rows.push(row);
  }
  if (format === "csv") {
    const escape = (value: string) => {
      const safe = /^[=+\-@\t\r\n]/.test(value.trimStart())
        ? "'" + value
        : value;
      return /[",\r\n]/.test(safe)
        ? '"' + safe.replaceAll('"', '""') + '"'
        : safe;
    };
    return new Blob(
      [
        "\ufeff",
        rows.map((row) => row.map(escape).join(",") + "\r\n").join(""),
      ],
      { type: "text/csv;charset=utf-8" },
    );
  }
  const book = utils.book_new();
  utils.book_append_sheet(book, utils.aoa_to_sheet(rows), "VeriForge");
  return new Blob(
    [write(book, { type: "array", bookType: "xlsx", compression: true })],
    {
      type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    },
  );
}

File: engine/exports.py
import csv
from collections.abc import Iterable
from dataclasses import fields
from typing import TextIO

from engine.models import Result, Status

EXPORT_COLUMNS = {
    "valid": ["email"],
    "corrected": ["original_email", "corrected_email", "reason"],
    "removed": ["original_email", "category", "reason", "domain"],
    "review": ["original_email", "final_email", "reason", "domain"],
    "clean": ["email"],
    "full": [field.name for field in fields(Result)],
}


def export_row(record: Result, kind: str) -> list | None:
    if kind not in EXPORT_COLUMNS:
        raise ValueError("Unknown export type")
    if kind == "clean":
        return [record.final_email] if record.status in (Status.VALID, Status.CORRECTED) else None
    if kind == "valid":
        return [record.final_email] if record.status == Status.VALID else None
    if kind == "full":
        return [getattr(record, key) for key in EXPORT_COLUMNS[kind]]
    if record.status.value.lower() != kind:
        return None
    return [
        getattr(record, "final_email" if key == "corrected_email" else key)
        for key in EXPORT_COLUMNS[kind]
    ]


def safe_csv_cell(value: object) -> str:
    text = str(value)
    # Quoting alone does not prevent formulas in spreadsheet programs.
    if text.lstrip().startswith(("=", "+", "-", "@", "\t", "\r", "\n")):
        return "'" + text
    return text


def write_csv(records: Iterable[Result], output: TextIO, kind: str) -> int:
    writer = csv.writer(output)
    writer.writerow(EXPORT_COLUMNS[kind])
    count = 0
    for record in records:
        row = export_row(record, kind)
        if row is not None:
            writer.writerow([safe_csv_cell(cell) for cell in row])
            count += 1
    return count
